### DISCUSSION TITLE
Problem with long RF waveforms ?

### DISCUSSION THREAD

**Discussion by tblazey** [JuliaHealth/KomaMRI.jl#178] [General]
Hi,

First off, KomaMRI is an amazing tool. Thanks for putting it out there for everyone to use!

I've been trying to use it to simulate some of the Pulseq sequences I've been developing and the results are not exactly what I've been expecting. In trying to understand what is going on, I noticed that the RF waveforms shown in KomaMRI are different what I see in the Pulseq MATLAB package or PyPulseq (see attached images).

It looks to me that KomaMRI is using a coarser time sampling than the other tools. Is this just in the plots, or do the simulations use this coarser sampling? If so, can I adjust the sampling somehow? Perhaps with the seq_Δt or RF_Δt option of Scanner? 

Thanks!

-Tyler

<img width="1371" alt="Koma_RF" src="https://github.com/cncastillo/KomaMRI.jl/assets/810152/3d545676-e3cb-4997-b9d1-c9d61427412f">

<img width="1235" alt="pyPulseq_RF" src="https://github.com/cncastillo/KomaMRI.jl/assets/810152/66e5a0dc-f156-4f40-83ee-72bbfd3a08b8">

<img width="1369" alt="Pulseq_RF" src="https://github.com/cncastillo/KomaMRI.jl/assets/810152/69b3683c-330b-452c-a9f0-757a624a484e">

[met_one_for_sim_seq.txt](https://github.com/cncastillo/KomaMRI.jl/files/12197224/met_one_for_sim_seq.txt)

**Comment by cncastillo** [comment]
Hi Tyler,

First of all, thanks for using Koma! I am very excited that people are trying to do stuff with it :smile:.

Regarding the plot, the number of samples for each RF object is capped at 100 for plotting, so that is why the train of sincs looks undersampled. The reason for this is that when you have a long sequence with a lot of RFs, the plot starts to become slow. So it is a plotting "problem".  Also, I never thought people would define an RF with 18690 samples! :open_mouth:.

The command `plot_seq(seq; max_rf_samples=Inf)` should show the sequence that is actually being loaded.  Maybe this should be the default behaviour. 
![newplot](https://github.com/cncastillo/KomaMRI.jl/assets/5957134/bafde7cd-4438-427a-b959-88030e3a7c01)

If you want to check what is actually being simulated you can inspect the sampling with `plot_seqd(seq)`. There you can see the effect that changing `Δt_rf` has (as your gradients are also arbitrary it will not have a huge effect, as it is also sampling there every `Δt`).

```julia
simParams = KomaMRICore.default_sim_params()
simParams["Δt_rf"] = 1 #1s, very exaggerated
plot_seqd(seq; simParams)
```
![newplot](https://github.com/cncastillo/KomaMRI.jl/assets/5957134/755754ae-256a-420b-9b60-cf49b5b26b27)
![newplot](https://github.com/cncastillo/KomaMRI.jl/assets/5957134/17b977d0-3b45-4ad9-a585-7fc63c02b5bf)
I saw the gradients are not very sharp in some areas (also saw this opening the file with MATLAB's Pulseq), maybe that could be causing the problem (?).

I think this is a great opportunity to document this better (the effect of `Δt` and `Δt_rf` in the simulation). Maybe @beorostica, can help. 

Hope this is helpful, please let me know if you need more help.
Cheers!

EDIT: Also I suggest updating Koma with `]up`, as we fixed some bugs in the last version that could be the cause of your problems.

**Comment by tblazey** [comment]
Thanks for the very quick reply!

I appreciate the pointer to max_rf_samples, I should have read the docs more carefully 🤦‍♂️.

Thanks again!

-Tyler
